src/tools/r2rdump/R2RReader.cs

```diff
@@ -65,25 +75,164 @@ public unsafe R2RReader(string filename)
                 Machine = peReader.PEHeaders.CoffHeader.Machine;
                 ImageBase = peReader.PEHeaders.PEHeader.ImageBase;
 
-                SectionHeader textSection;
-                int nSections = peReader.PEHeaders.CoffHeader.NumberOfSections;
-                for (int i = 0; i < nSections; i++)
-                {
-                    SectionHeader section = peReader.PEHeaders.SectionHeaders[i];
-                    if (section.Name.Equals(".text"))
-                    {
-                        textSection = section;
-                    }
-                }
-
+                // initialize the R2RHeader
                 DirectoryEntry r2rHeaderDirectory = peReader.PEHeaders.CorHeader.ManagedNativeHeaderDirectory;
-                int r2rHeaderOffset = r2rHeaderDirectory.RelativeVirtualAddress - textSection.VirtualAddress + textSection.PointerToRawData;
-                R2RHeader = new R2RHeader(_image, (uint)r2rHeaderDirectory.RelativeVirtualAddress, r2rHeaderOffset);
+                int r2rHeaderOffset = GetOffset(r2rHeaderDirectory.RelativeVirtualAddress);
+                R2RHeader = new R2RHeader(_image, r2rHeaderDirectory.RelativeVirtualAddress, r2rHeaderOffset);
                 if (r2rHeaderDirectory.Size != R2RHeader.Size)
                 {
                     throw new System.BadImageFormatException("The calculated size of the R2RHeader doesn't match the size saved in the ManagedNativeHeaderDirectory");
                 }
+
+                if (peReader.HasMetadata)
+                {
+                    var mdReader = peReader.GetMetadataReader();
+
+                    // initialize R2RMethods with method signatures from MethodDefHandle, and runtime function indices from MethodDefEntryPoints
+                    int i = 0;
+                    int methodDefEntryPointsRVA = R2RHeader.Sections[R2RSection.SectionType.READYTORUN_SECTION_METHODDEF_ENTRYPOINTS].RelativeVirtualAddress;
+                    int methodDefEntryPointsOffset = GetOffset(methodDefEntryPointsRVA);
+                    NativeArray methodEntryPoints = new NativeArray(_image, (uint)methodDefEntryPointsOffset);
+                    uint nMethods = methodEntryPoints.GetCount();
+                    R2RMethods = new R2RMethod[nMethods];
+                    for (uint rid = 1; rid <= nMethods; rid++)
+                    {
+                        uint offset = 0;
+                        if (methodEntryPoints.TryGetAt(_image, rid - 1, ref offset))
+                        {
+                            MethodDefinitionHandle methodDefHandle = MetadataTokens.MethodDefinitionHandle((int)rid);
```

Comment:
nit: You can probably move this call into the R2RMethod constructor. It would be nice to keep the list of arguments as succinct as possible.